streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableTransformValuesTest.java

```diff
@@ -0,0 +1,553 @@
+                        new ExclamationValueTransformerSupplier(STORE_NAME, OTHER_STORE_NAME),
+                        STORE_NAME, OTHER_STORE_NAME)
+                .toStream()
+                .process(capture);
+
+        driver = new TopologyTestDriver(builder.build(), props());
+
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "a", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "B", "b", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "D", null, 0L));
+
+        assertThat(output(), hasItems("A:A->a!", "B:B->b!", "D:null"));
+    }
+
+    @Test
+    public void shouldTransformValuesWithKeyAndMaterialize() {
+        builder.addStateStore(storeBuilder(STORE_NAME))
+                .table(INPUT_TOPIC, CONSUMED)
+                .transformValues(
+                        new ExclamationValueTransformerSupplier(STORE_NAME, QUERYABLE_NAME),
+                        Materialized.<String, String, KeyValueStore<Bytes, byte[]>>as(QUERYABLE_NAME)
+                                .withKeySerde(Serdes.String())
+                                .withValueSerde(Serdes.String()),
+                        STORE_NAME)
+                .toStream()
+                .process(capture);
+
+        driver = new TopologyTestDriver(builder.build(), props());
+
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "a", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "B", "b", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "C", null, 0L));
+
+        assertThat(output(), hasItems("A:A->a!", "B:B->b!", "C:null"));
+
+        final KeyValueStore<String, String> keyValueStore = driver.getKeyValueStore(QUERYABLE_NAME);
+        assertThat(keyValueStore.get("A"), is("A->a!"));
+        assertThat(keyValueStore.get("B"), is("B->b!"));
+        assertThat(keyValueStore.get("C"), is(nullValue()));
+    }
+
+    @Test
+    public void shouldCalculateCorrectOldValuesIfMaterializedEvenIfStateful() {
+
+        builder.addStateStore(storeBuilder(STORE_NAME))
+                .table(INPUT_TOPIC, CONSUMED)
+                .transformValues(
+                        new StatefulTransformerSupplier(),
+                        Materialized.<String, Integer, KeyValueStore<Bytes, byte[]>>as(QUERYABLE_NAME)
+                                .withKeySerde(Serdes.String())
+                                .withValueSerde(Serdes.Integer()))
+                .groupBy(toForceSendingOfOldValues(), Serialized.with(Serdes.String(), Serdes.Integer()))
+                .reduce(MockReducer.INTEGER_ADDER, MockReducer.INTEGER_SUBTRACTOR)
+                .mapValues(mapBackToStrings())
+                .toStream()
+                .process(capture);
+
+        driver = new TopologyTestDriver(builder.build(), props());
+
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "a", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "aa", 0L));
```

Comment:
maybe use "a", "b", "c" as values, as the transformer counts the number of calls to `process` (for better distinction with next test)

Reply:
yer, or maybe just 'ignored'